### PR TITLE
Fix LightningDataModule.load_from_checkpoint to restore subclass and hyperparameters

### LINKED ISSUE #21477: `LightningDataModule.load_from_checkpoint` does not restore subclass from `datamodule_hyper_parameters`
### Bug description

`LightningDataModule.load_from_checkpoint` fails to instantiate the **specific DataModule subclass**  whose class path is stored in the checkpoint under `datamodule_hyper_parameters`  in `_class_path`.

While `LightningModule.load_from_checkpoint` correctly restores the concrete subclass, `LightningDataModule.load_from_checkpoint` instead returns a plain `LightningDataModule` instance and silently ignoring all the subclass's hyperparameters. No warning is raised. 

### What version are you seeing the problem on?

master

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
from pathlib import Path
from unittest import mock

import torch
from jsonargparse import lazy_instance
from lightning import LightningDataModule
from lightning.pytorch.cli import (
    LightningCLI,
    LRSchedulerCallable,
    OptimizerCallable,
)
from lightning.pytorch.demos.boring_classes import (
    BoringDataModule,
    BoringModel,
)


class TestDataSaveHparams(BoringDataModule):
    def __init__(self, batch_size: int = 32, num_workers: int = 4):
        super().__init__()
        self.save_hyperparameters()
        self.batch_size = batch_size
        self.num_workers = num_workers


class TestModelSaveHparams(BoringModel):
    def __init__(
        self,
        optimizer: OptimizerCallable = torch.optim.Adam,
        scheduler: LRSchedulerCallable = torch.optim.lr_scheduler.ConstantLR,
        activation: torch.nn.Module = lazy_instance(
            torch.nn.LeakyReLU, negative_slope=0.05
        ),
    ):
        super().__init__()
        self.save_hyperparameters()
        self.optimizer = optimizer
        self.scheduler = scheduler
        self.activation = activation

    def configure_optimizers(self):
        optimizer = self.optimizer(self.parameters())
        scheduler = self.scheduler(optimizer)
        return {"optimizer": optimizer, "lr_scheduler": scheduler}


def test_lightning_cli_load_from_checkpoint_for_datamodule(cleandir):
    with mock.patch(
        "sys.argv",
        [
            "any.py",
            "--trainer.max_epochs=1",
            "--model=TestModelSaveHparams",
            "--data=TestDataSaveHparams",
        ],
    ):
        cli = LightningCLI(
            run=False,
            auto_configure_optimizers=False,
        )
    cli.trainer.fit(cli.model, datamodule=cli.datamodule)

    expected_keys = ["_class_path", "_instantiator", "batch_size", "num_workers"]
    expected_instantiator = "lightning.pytorch.cli.instantiate_module"
    expected_class_path = f"{__name__}.TestDataSaveHparams"
    expected_batch_size = 32
    expected_num_workers = 4

    checkpoint_path = next(
        Path(cli.trainer.log_dir, "checkpoints").glob("*.ckpt"), None
    )
    assert checkpoint_path.is_file()
    loaded_checkpoint = torch.load(checkpoint_path, weights_only=True)
    dm_hparams = loaded_checkpoint[LightningDataModule.CHECKPOINT_HYPER_PARAMS_KEY]

    assert sorted(dm_hparams.keys()) == expected_keys
    assert dm_hparams["_instantiator"] == expected_instantiator
    assert dm_hparams["_class_path"] == expected_class_path
    assert dm_hparams["batch_size"] == expected_batch_size
    assert dm_hparams["num_workers"] == expected_num_workers

    dm = LightningDataModule.load_from_checkpoint(checkpoint_path)
    assert isinstance(dm, TestDataSaveHparams)
    dm.batch_size == expected_batch_size
    dm.num_workers == expected_num_workers


test_lightning_cli_load_from_checkpoint_for_datamodule("")


Note: Certain portion of the above code is taken from https://github.com/Lightning-AI/pytorch-lightning/blob/master/tests/tests_pytorch/test_cli.py to reproduce the bug with minimalistic example
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### Root cause analysis

`load_from_checkpoint` internally calls `_load_state` from `src/lightning/pytorch/core/saving.py`

During instantiation in `load_state` function, the following logic filters constructor arguments: 
https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/core/saving.py#L163-L165

```python
if not cls_spec.varkw:
    # filter kwargs according to class init unless it allows any argument via kwargs
    _cls_kwargs = {k: v for k, v in _cls_kwargs.items() if k in cls_init_args_name}
```

This works for `LightningModule` because it defines the kwargs and hence the above condition is bypassed for it:

```python
class LightningModule(...):
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        super().__init__(*args, **kwargs)
```

However, `LightningDataModule` **does not accept `*args, **kwargs`**, so:

* `cls_spec.varkw` is `False`
* All subclass constructor arguments  are filtered out
* The instantiator falls back to constructing the base `LightningDataModule` (https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/core/saving.py#L167)

cc @ethanwharris @lantiga

### PR DESCRIPTION
**Pull request by aditya0by0**
## What does this PR do?

`LightningDataModule.load_from_checkpoint` currently fails to restore the concrete DataModule subclass and its hyperparameters from the checkpoint. This happens because `LightningDataModule.__init__` does not accept `*args` and `**kwargs`, causing the instantiator to filter out subclass-specific arguments.

This PR updates `LightningDataModule.__init__` to accept `*args` and `**kwargs`, aligning its behavior with `LightningModule`. After this change, `load_from_checkpoint` correctly instantiates the subclass and restores all hyperparameters.

Fixes #21477

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21478.org.readthedocs.build/en/21478/

### CODE REVIEW

**Comment by GdoongMathew** [comment]
Hi @aditya0by0 , I just left a my comment in the original issue. Please feel free to let me know if you have any feedback or a different perspective - happy to discuss further 🙂

File: src/lightning/pytorch/core/datamodule.py [hunk 1]
```diff
@@ -80,8 +80,8 @@ def teardown(self):
     CHECKPOINT_HYPER_PARAMS_NAME = "datamodule_hparams_name"
     CHECKPOINT_HYPER_PARAMS_TYPE = "datamodule_hparams_type"
 
-    def __init__(self) -> None:
-        super().__init__()
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
```
Comment (GdoongMathew) [inline src/lightning/pytorch/core/datamodule.py]:
```suggestion
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        super().__init__()
```

Since none of it super classes, (`DataHooks` and `HyperparameterMixin`) accepts additional arguments, that should be removed.

File: tests/tests_pytorch/test_cli.py [hunk 2]
```diff
@@ -1075,6 +1075,36 @@ def test_lightning_cli_load_from_checkpoint_dependency_injection_subclass_mode(c
     assert isinstance(lr_scheduler, torch.optim.lr_scheduler.ConstantLR)
 
 
+def test_lightning_cli_load_from_checkpoint_for_datamodule(cleandir):
```
Comment (GdoongMathew) [inline tests/tests_pytorch/test_cli.py]: I think this test could be added within the other test `test_lightning_cli_load_from_checkpoint_dependency_injection_subclass_mode`

**Review by GdoongMathew — CHANGES_REQUESTED** [review]

**Comment by aditya0by0** [comment]
@GdoongMathew, I have implemented the suggested changes.
